File: app/src/main/java/igrek/songbook/layout/contact/ContactLayoutController.java
package igrek.songbook.layout.contact;

import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.R;
import igrek.songbook.activity.ActivityController;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.info.UiInfoService;
import igrek.songbook.info.UiResourceService;
import igrek.songbook.info.errorcheck.SafeClickListener;
import igrek.songbook.info.logger.Logger;
import igrek.songbook.info.logger.LoggerFactory;
import igrek.songbook.layout.LayoutController;
import igrek.songbook.layout.LayoutState;
import igrek.songbook.layout.MainLayout;
import igrek.songbook.layout.navigation.NavigationMenuController;
import igrek.songbook.layout.view.ButtonClickEffect;

public class ContactLayoutController implements MainLayout {
	
	@Inject
	Lazy<ActivityController> activityController;
	@Inject
	LayoutController layoutController;
	@Inject
	UiInfoService uiInfoService;
	@Inject
	UiResourceService uiResourceService;
	@Inject
	AppCompatActivity activity;
	@Inject
	NavigationMenuController navigationMenuController;
	@Inject
	SendFeedbackService sendFeedbackService;
	
	private Logger logger = LoggerFactory.getLogger();
	private EditText contactMessageEdit;
	private EditText contactAuthorEdit;
	
	public ContactLayoutController() {
		DaggerIoc.getFactoryComponent().inject(this);
	}
	
	@Override
	public void showLayout(View layout) {
		// Toolbar
		Toolbar toolbar1 = layout.findViewById(R.id.toolbar1);
		activity.setSupportActionBar(toolbar1);
		ActionBar actionBar = activity.getSupportActionBar();
		if (actionBar != null) {
			actionBar.setDisplayHomeAsUpEnabled(false);
			actionBar.setDisplayShowHomeEnabled(false);
		}
		// navigation menu button
		ImageButton navMenuButton = layout.findViewById(R.id.navMenuButton);
		navMenuButton.setOnClickListener((v) -> navigationMenuController.navDrawerShow());
		ButtonClickEffect.addClickEffect(navMenuButton);
		
		contactMessageEdit = layout.findViewById(R.id.contactMessageEdit);
		contactAuthorEdit = layout.findViewById(R.id.contactAuthorEdit);
		Button contactSendButton = layout.findViewById(R.id.contactSendButton);
		contactSendButton.setOnClickListener(new SafeClickListener() {
			@Override
			public void onClick() {
				sendContactMessage();
			}
		});
	}
	
	@Override
	public LayoutState getLayoutState() {
		return LayoutState.CONTACT;
	}
	
	@Override
	public int getLayoutResourceId() {
		return R.layout.contact;
	}
	
	@Override
	public void onBackClicked() {
		layoutController.showSongTree();
	}
	
	private void sendContactMessage() {
		String message = contactMessageEdit.getText().toString();
		String author = contactAuthorEdit.getText().toString();
		if (message == null || message.isEmpty()) {
			uiInfoService.showToast("Message field can't be empty.");
			return;
		}
		sendFeedbackService.sendFeedback(message, author);
	}
}

File: app/src/main/java/igrek/songbook/send/ContactLayoutController.kt
package igrek.songbook.send

import android.view.View
import android.widget.Button
import android.widget.EditText
import igrek.songbook.R
import igrek.songbook.info.UiInfoService
import igrek.songbook.info.UiResourceService
import igrek.songbook.info.analytics.AnalyticsLogger
import igrek.songbook.info.errorcheck.SafeClickListener
import igrek.songbook.inject.LazyExtractor
import igrek.songbook.inject.LazyInject
import igrek.songbook.inject.appFactory
import igrek.songbook.layout.LayoutController
import igrek.songbook.layout.MainLayout
import igrek.songbook.layout.dialog.ConfirmDialogBuilder
import igrek.songbook.system.SoftKeyboardService

class ContactLayoutController(
    layoutController: LazyInject<LayoutController> = appFactory.layoutController,
    uiInfoService: LazyInject<UiInfoService> = appFactory.uiInfoService,
    uiResourceService: LazyInject<UiResourceService> = appFactory.uiResourceService,
    sendMessageService: LazyInject<SendMessageService> = appFactory.sendMessageService,
    softKeyboardService: LazyInject<SoftKeyboardService> = appFactory.softKeyboardService,
) : MainLayout {
    private val layoutController by LazyExtractor(layoutController)
    private val uiInfoService by LazyExtractor(uiInfoService)
    private val uiResourceService by LazyExtractor(uiResourceService)
    private val sendMessageService by LazyExtractor(sendMessageService)
    private val softKeyboardService by LazyExtractor(softKeyboardService)

    private var contactSubjectEdit: EditText? = null
    private var contactMessageEdit: EditText? = null
    private var contactAuthorEdit: EditText? = null

    override fun showLayout(layout: View) {
        contactSubjectEdit = layout.findViewById(R.id.contactSubjectEdit)
        contactMessageEdit = layout.findViewById(R.id.contactMessageEdit)
        contactAuthorEdit = layout.findViewById(R.id.contactAuthorEdit)

        val contactSendButton = layout.findViewById<Button>(R.id.contactSendButton)
        contactSendButton.setOnClickListener(SafeClickListener {
            sendContactMessage()
        })
    }

    override fun getLayoutResourceId(): Int {
        return R.layout.screen_contact
    }

    override fun onBackClicked() {
        layoutController.showPreviousLayoutOrQuit()
    }

    override fun onLayoutExit() {
        softKeyboardService.hideSoftKeyboard()
    }

    private fun sendContactMessage() {
        val message = contactMessageEdit!!.text.toString()
        val author = contactAuthorEdit!!.text.toString()
        val subject = contactSubjectEdit!!.text.toString()
        if (message.isEmpty()) {
            uiInfoService.showToast(uiResourceService.resString(R.string.contact_message_field_empty))
            return
        }
        ConfirmDialogBuilder().confirmAction(R.string.confirm_send_contact) {
            sendMessageService.sendContactMessage(
                message, origin = MessageOrigin.CONTACT_MESSAGE,
                author = author, subject = subject
            )
            AnalyticsLogger().logEventContactMessageSent()
        }
    }
}
